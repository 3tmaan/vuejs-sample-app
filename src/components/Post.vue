<template>
  <div class="card pl-3 pr-8 py-4 mb-4">
    <h3 class="text-base">Post {{ post.id }}</h3>
    <p class="text-xs mt-2 mr-8">{{ post.title }}</p>
    <div class="absolute right-0 top-0 mr-2 h-full flex flex-col justify-center">
      <button class="btn" ref="buttonUp" type="button" @click="moveUp(index)" v-if="index !== 0">
        <svg class="w-3" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M0 16.67l2.829 2.83 9.175-9.339 9.167 9.339 2.829-2.83-11.996-12.17z"/></svg>
      </button>
      <button class="btn" ref="buttonDown" type="button" @click="moveDown(index)" v-if="index !== posts.length-1">
        <svg class="w-3" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M0 7.33l2.829-2.83 9.175 9.339 9.167-9.339 2.829 2.83-11.996 12.17z"/></svg>
      </button>
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex';

export default {
  name: 'Post',
  props: {
    index: {
      type: Number,
      required: true,
    }
  },
  computed: {
    ...mapGetters(["posts"]),
    post() {
      return this.posts[this.index]
    }
  },
  methods: {
    moveUp(index) {
      this.$store.commit('MOVE_UP', {index});
    },
    moveDown(index) {
      this.$store.commit('MOVE_DOWN', {index});
    }
  }
}
</script>

<style scoped>
  .btn {
    @apply p-1 my-3 outline-none
  }
</style>