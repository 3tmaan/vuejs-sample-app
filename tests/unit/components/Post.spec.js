import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex'
import Post from '@/components/Post.vue';
import mockedPosts from '../../../fixtures/posts';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Post.vue', () => {
    let store;
    let mutations;
    let getters;
    let options;

    beforeEach(() => {
        mutations =  {
            MOVE_UP: jest.fn(),
            MOVE_DOWN: jest.fn(),
        }
        getters = {
            posts: () => mockedPosts
        }
        store = new Vuex.Store({
            mutations,
            getters
        })
        options = {
            store,
            localVue,
            propsData: { index: 1 }
        }
    });

    it('renders correctly', () => {
        const wrapper = shallowMount(Post, options);

        expect(wrapper).toMatchSnapshot();
    });

    it('calls store mutations method "MOVE_UP" when "up arrow" button is clicked', () => {
        const wrapper = shallowMount(Post, options);
        const buttonUp = wrapper.find({ ref: 'buttonUp' });

        buttonUp.trigger('click');
        expect(mutations.MOVE_UP).toHaveBeenCalled();
    });

    it('calls store mutations method "MOVE_DOWN" when "down arrow" button is clicked', () => {
        const wrapper = shallowMount(Post, options);
        const buttonDown = wrapper.find({ ref: 'buttonDown' });

        buttonDown.trigger('click');
        expect(mutations.MOVE_DOWN).toHaveBeenCalled();
    });
});
