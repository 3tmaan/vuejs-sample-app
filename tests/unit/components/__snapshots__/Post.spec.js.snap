// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post.vue renders correctly 1`] = `
<div class="card pl-3 pr-8 py-4 mb-4">
  <h3 class="text-base">Post 1</h3>
  <p class="text-xs mt-2 mr-8">sunt aut facere repellat provident occaecati excepturi optio reprehenderit</p>
  <div class="absolute right-0 top-0 mr-2 h-full flex flex-col justify-center"><button type="button" class="btn"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="w-3">
        <path d="M0 16.67l2.829 2.83 9.175-9.339 9.167 9.339 2.829-2.83-11.996-12.17z"></path>
      </svg></button> <button type="button" class="btn"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="w-3">
        <path d="M0 7.33l2.829-2.83 9.175 9.339 9.167-9.339 2.829 2.83-11.996 12.17z"></path>
      </svg></button></div>
</div>
`;
